### ISSUE TITLE
ProjectBuilder ProviderFactory support

### ISSUE THREAD

**Issue by Sineaggi** [gradle/gradle#20690]
### Expected Behavior
```java
Project project = new ProjectBuilder()
    .withEnvironmentVariable(key, value)
    .withProperty(key, value)
    .withGradleProperty(key, value)
    .build()
```

I would not expect these to actually set environment variables or system properties behind the scenes. These settings should only be made available to plugins using the ProviderFactory (`project.providers`) interface.

### Current Behavior
We need to use something like [JUnit Pioneer](https://junit-pioneer.org/)'s [SetEnvironmentVariable](https://junit-pioneer.org/docs/environment-variables/) to modify environment variables which would not be thread-safe and require illegal reflection to modify.

### Context

We're working on gradle plugins that can be configured based on the environment they run in. This means we'd like to load environment variables or system properties to modify plugin behavior. Testing these behaviors needs some way to either inject or override the environment variable loading code. We'd like plugins to be able to load environment variables from the `ProjectProvider` interface, and then override them in the `ProjectBuilder` api.

**Comment by hegyibalint** [comment]
You might want to consider using `TestKit`?
https://docs.gradle.org/current/userguide/test_kit.html

**Comment by tridium-gward** [comment]
I can't speak for @Sineaggi , but at least in our experience `TestKit` is too slow to be suitable for unit testing -- we use `ProjectBuilder` for unit tests and `TestKit` for functional/integration tests. 

Another workaround for this issue we've used successfully is to define an extension that wraps all the property configurations we use. For example:

```kotlin
abstract class BuildConfiguration @Inject constructor(
  objectFactory: ObjectFactory,
  providerFactory: ProviderFactory
)  {
  @get:Input
  @get:Optional
  val nodeHomeGradleProperty: Property<String> = objectFactory.property<String>().convention(
    providerFactory.gradleProperty("nodeHome").forUseAtConfigurationTime()
  )
}
```
(I think `forUseAtConfigurationTime` is deprecated, but we're still on 7.3)

This extension gets added to the root project by a plugin separate from any plugins that rely on it for their `apply` method.  This allows you to add the extension in your test and set the value of the property directly before applying any additional plugins, while relying on `providerFactory.gradleProperty()` in real builds (which you can test using `TestKit` for an integration test)

**Comment by Sineaggi** [comment]
Yeah so for us TestKit is slow and more useful for end-to-end testing.
We want to validate that our tasks and extension objects are configured correctly. This is much easier with the ProjectBuilder.

**Comment by jbartok** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not in focus. It might take a while before it gets implemented.

***

Relates to
* #17638

**Comment by wakingrufus** [comment]
+1 to this. unit tests using ProjectBuilder are still useful to test plugin/configuration code (not task execution) in a more performant way than a full testkit build

**Comment by wakingrufus** [comment]
A wrinkle in implementing this will be that Gradle objects can also inject ProviderFactory, not just use `project.providers` and I believe the ProjectBuilder doesn't support services or injection at all currently
